### ISSUE TITLE
Problem casting time from json input

### ISSUE THREAD

**Issue by ethack** [brimdata/zed#3987]
I'm having a problem using the Zeek [shaper.zed](https://github.com/brimdata/zed/blob/1f3656fb0a5e933b8bca7a86eba03f76a43f58c1/zeek/shaper.zed) while parsing JSON files. Specifically, the `ts` field doesn't seem to be parsing correctly.

For a minimal example, I'm recreating the commands from [this comment](https://github.com/brimdata/zed/issues/808#issuecomment-922133431). The following command returned a correct date for a Unix timestamp conversion in the original comment, but with v1.1.0 it is giving an incorrect date. (I'm not worried about the floating point rounding issues from the original post.)
```
$ zq -version 
Version: v1.1.0

$ echo '{"mytime": 1617490638.707690}' | zq 'timeval:=time(mytime)' -
{mytime:1.61749063870769e+09,timeval:1970-01-01T00:00:01.617490638Z}
```

When using the Zeek TSV example, v1.1.0 still returns the same correct result as the original comment.
```
$ cat mytime.tsv      
#separator \x09
#set_separator  ,
#empty_field    (empty)
#unset_field    -
#fields timeval
#types  time
1617490638.707690

$ zq -z mytime.tsv                                                          
{timeval:2021-04-03T22:57:18.70769Z}
```

As I mentioned, I'm trying to use the [shaper.zed](https://github.com/brimdata/zed/blob/1f3656fb0a5e933b8bca7a86eba03f76a43f58c1/zeek/shaper.zed) file. Here's an example that's closer to my actual usage that also shows the problem:

```
$ echo '{"_path":"conn","ts":1657238339.882122}' | zq -z -I shaper.zed -
{_path:"conn",ts:1970-01-01T00:00:01.657238339Z,uid:null(string),id:null(conn_id={orig_h:ip,orig_p:port=uint16,resp_h:ip,resp_p:port}),proto:null(zenum=string),service:null(string),duration:null(duration),orig_bytes:null(uint64),resp_bytes:null(uint64),conn_state:null(string),local_orig:null(bool),local_resp:null(bool),missed_bytes:null(uint64),history:null(string),orig_pkts:null(uint64),orig_ip_bytes:null(uint64),resp_pkts:null(uint64),resp_ip_bytes:null(uint64),tunnel_parents:null(|[string]|),_write_ts:null(time)}(=conn)
```

Has something changed with the timestamp conversion so my usage is wrong?

**Comment by philrz** [comment]
@ethack: Yes, something did change. Thanks for reporting. Alas, like many things Zeek-related, it gets a little complex, so please bear with me. :smile:

To get right to your question, the change came in #3809. In Zed, t[he `time` type is stored internally as nanoseconds from epoch](3809). Before the change in #3809, the `time()` cast treated numeric values like `1617490638.707690` as _seconds_ since epoch. That meant that to store them in the proper Zed `time` type it would have to "scale" the value to the correct range by multiplying it by 1-billion. As the PR notes indicate, this had the potential to create confusion because if a user later wanted to do the opposite (convert a `time` type back to a numeric type), there's no clear way to know if the user's intent would be to scale it back to seconds or to preserve its nanosecond nature (since not all times in all data sources started life as epoch seconds), so the user needs to make an informed scaling decision to get correct data. The change in #3809 resolves this by making the scaling decision happen up front when first going from number to `time`. Unfortunately since Zeek numeric JSON time is a log format that uses this "seconds since epoch" notation, that change in behavior broke the reference shaper as you experienced.

The reason why the Zeek TSV format has a similar-looking numeric time but doesn't have this problem is because Zed has purpose-built Zeek TSV reader/writer code, so it's aware of how to turn a Zeek `time` type into a Zed `time` type and back again without any manual scaling. There's actually many other advantages of Zeek TSV over Zeek JSON, so if you weren't attached to JSON, it's a great choice of Zeek log format. But I assume you've got other tooling that needs JSON.

Once again focusing on the narrow example, the quick fix to get that example line working again would be to multiply the numeric value by 1-billion before doing the `time()` cast.

```
$ echo '{"mytime": 1617490638.707690}' | zq -Z 'timeval:=time(mytime*1000000000)' -
{
    mytime: 1.61749063870769e+09,
    timeval: 2021-04-03T22:57:18.707689984Z
}
```

However, this approach becomes a bit of a pain if you want to make sure all the numeric JSON fields that started life as the `time` type in Zeek get the same treatment, as there's others hiding in there (e.g., the [Kerberos `till` field](https://docs.zeek.org/en/master/scripts/base/protocols/krb/main.zeek.html#id2) is one such example), so enumerating them all by name and log type and making sure they get this conversion in the shaper would be a little ugly. In thinking this over, it might be great if Zed had some way to add custom logic to the [`cast()`](https://zed.brimdata.io/docs/next/language/functions/cast/) part of shaping, e.g., something like "when this shaper casts any field via `time()`, scale the value by multiplying by a billion first". But this doesn't currently exist in the language.

While I don't love asking folks to change their data to fit the tooling, if you're not attached to having the times in the Zeek JSON logs stay as float seconds, the quickest and cleanest fix would probably be to change the output format of the Zeek `time` fields. I'm not sure if you picked up on the comment [here](https://github.com/brimdata/zed/blob/main/zeek/Reading-Zeek-Log-Formats.md#zeek-ndjson), but when writing these Zeek reference materials for Zed we actually used the logs output by the [json-streaming-logs](https://github.com/corelight/json-streaming-logs) package, and those happen to output Zeek time values as ISO8601 strings, which the Zed `time()` cast handles just fine. Most of the Zeek JSON users we've bumped into use that package because they were accustomed to sending their logs to tools like Elastic and Splunk that generally worked better "out of the box" with the ISO timestamps, so we opted to go the same route in our testing/materials. I'm not sure if that package is at all tempting to you, but that's one option.

Since you've got the JSON numbers as epoch seconds, I expect you're using the JSON output mode of Zeek's built-in [ASCII log writer](https://docs.zeek.org/en/master/scripts/base/frameworks/logging/writers/ascii.zeek.html), i.e., enabled via:

```
redef LogAscii::use_json = T;
```

If that's the case and you're ok with using the ISO timestamp strings instead of the epoch seconds as a number, you could add this additional line of Zeek config:

```
redef LogAscii::json_timestamps = JSON::TS_ISO8601;
```

Finally, if you have a strong preference to stick with the epoch seconds as a JSON number but scale all the `time` fields in the shaper, I think all the conditional logic exists in Zed to add that to the shaper, and I'd be game to give it a shot. Since it'd take some effort and may look a bit clunky in the end, I figured I'd double check first if the other options were workable. :smile: 

Please do let me know the direction you might want to go with this and/or if you have any additional questions. Thanks for your interest in Zed!

**Comment by ethack** [comment]
Thank you for the explanation and the workaround!

I can appreciate the benefits of ISO8601 timestamps. I work with Zeek systems managed by others and don't necessarily have access to change the Zeek settings directly. I could likely convince them to make changes eventually, but the short term path of least resistance for me will be supporting whichever timestamp variation is used.

Sidenote: In this case the data is coming from Security Onion. I'm a bit surprised SO doesn't use ISO8601 since the logs get parsed into Elastic. But as near as I can tell, `use_json` is the [default setting](https://github.com/Security-Onion-Solutions/securityonion/blob/65165e52f4c77a39833a1e89c00ea0b799bf1257/pillar/zeek/init.sls#L54). ISO8601 used to be a default setting, but looks like it got removed/reverted [here](https://github.com/Security-Onion-Solutions/securityonion/commit/37929dbd7d09d0da03919541c792eb0cefa25fde#diff-308ae435fe6ba11692a8dea86d60fd8b7ce937bffea344e3716b77058bdac650L55). I'm not sure what the reasoning was.

I'm already having to massage the logs I have because `_path` isn't set, I'll add in the `ts` conversion before applying the shaper.
```
zq 'put _path:="conn", ts:=time(ts*1000000000)' conn.00:00:00-01:00:00.log.gz | zq -I /host/root/zed/shaper.zed -
```
I'll keep an eye out for other time fields while I go.

I'm really enjoying the power of the Zed language since the v1 changes.

**Comment by philrz** [comment]
@ethack: Thanks for the context on your use case. That's interesting about Security Onion and their back & forth with ISO timestamps.

I'm glad to hear that you're ok with the workaround for now, so I'll go ahead and close this issue. When I have some spare cycles for bonus Zed hacking, I may try my hand at writing the conditional logic so the reference shaper could handle either type of timestamp automatically for all known Zeek `time` fields. If/when that happens, I'll post an update here. Also, if you've not already found your way onto it, the [Brim community Slack](https://www.brimdata.io/join-slack/) is also available, and you can ping me there (user "phil") any time.

**Comment by ethack** [comment]
Following up on the shaper idea, this is what I came up with. It's a line to conditionally change the value of `ts` before any shaping gets done.

```
// type definitions above
...

// scale ts to nanoseconds by detecting if:
// 1) it's a number (not a string or time)
// 2) is less than the expected nanosecond range
put ts:=(!(typeof(ts) in [<string>,<time>]) and ts < 1000000000000000000 ? ts * 1000000000 : ts) |

// This is from the default shaper.zed
yield nest_dotted(this) | switch (
  case _path=="ssl" and has(ssl_history) => yield shape(<ssl_4_1_0>)
  case _path=="x509" and has(fingerprint) => yield shape(<x509_4_1_0>)
  default => yield shape(schemas[_path])
) 
```
